### PR TITLE
Enabling the publish of CLI NuPkg to the unified Transport feed.

### PR DESCRIPTION
**Pull request by ghost**
The upload was tested on my dev box. Here is the blob location:
https://dotnetfeed.blob.core.windows.net/jbeisner/packages/microsoft.dotnet.cli.utils/

--------------------------------
Step 2 for: https://github.com/dotnet/cli/issues/7793

Notes:
build_projects/dotnet-cli-build/dotnet-cli-build.csproj:
For now, we will download the 'Microsoft.DotNet.Build.Tasks.Feed' package by placing a reference in 'dotnet-cli-build.csproj'; this is soon to be replaced by a "toolset" faux project per TaGoo.

build/publish/PublishNupkgToTransportFeed.targets:
For now, the 'PublishNupkgToTransportFeed' target will be a 'VSO' style step similar to: "Publish Windows Installer Nupkg to VS Feed" and "Publish MSBuild Extensions NuPkg to VS Feed" - controlled by "PUBLISH_NUPKG_TO_TRANSPORT_FEED" Boolean.

build/AzureInfo.props:
The value for 'TRANSPORTFEED_STORAGE_KEY' will be passed into the build via a secret parameter from VSO. The key will be stored in our Azure KeyVault.

### CODE REVIEW

File: build/Publish.targets [hunk 1]
```diff
@@ -33,7 +35,7 @@
                      Condition=" '$(IslinuxPortable)' != 'true' "/>
       <ForPublishing Include="%(GenerateArchivesInputsOutputs.Outputs)"
                      Condition=" '$(IsLinuxDistroSpecific)' != 'true' "/>
-      <ForPublishing Include="$(PackagesDirectory)/Microsoft*.nupkg"
+      <ForPublishing Include="$(PackagesDirectory)/Microsoft.DotNet.Cli.Utils.*.nupkg"
```
Comment (livarcocc) [inline build/Publish.targets]: This should include the resolver package as well.

File: build/publish/PublishNupkgToTransportFeed.targets [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    <RelativePath>packages</RelativePath>
+  </PropertyGroup>
+
+  <Target Name="PublishNupkgToTransportFeed" Condition=" '$(PUBLISH_NUPKG_TO_TRANSPORT_FEED)' == 'true' AND '$(OS)' == 'Windows_NT' And '$(Architecture)' == 'x64' " >
+    <Error Condition="'$(TransportFeedContainerName)' == ''" Text="Missing property TransportFeedContainerName." />
+    <Error Condition="'$(TransportFeedCloudDropAccountName)' == ''" Text="Missing property TransportFeedCloudDropAccountName." />
+    <Error Condition="'$(TransportFeedCloudDropAccessToken)' == ''" Text="Missing property TransportFeedCloudDropAccessToken." />
+
+    <Message Text="Publish to $(TransportFeedContainerName) started" />
+    <PushToBlobFeed AccountKey="$(TransportFeedCloudDropAccessToken)"
```
Comment (livarcocc) [inline build/publish/PublishNupkgToTransportFeed.targets]: How is this different than just uploading something to blob storage?

Comment (johnbeisner) [reply build/publish/PublishNupkgToTransportFeed.targets]: I gets registered in the index.json

File: build/Publish.targets (hunk 1 above)
Comment (johnbeisner) [reply build/Publish.targets]: Thanks for catching this.

**Comment by johnbeisner** [comment]
@jcagme 
@karajas 
Can I get a review on this PR please.

**Comment by johnbeisner** [comment]
@jcagme
@karajas
@chcosta 
Can I get a review on this PR please.

File: build/Publish.targets [hunk 3]
```diff
@@ -33,11 +35,13 @@
                      Condition=" '$(IslinuxPortable)' != 'true' "/>
       <ForPublishing Include="%(GenerateArchivesInputsOutputs.Outputs)"
                      Condition=" '$(IsLinuxDistroSpecific)' != 'true' "/>
-      <ForPublishing Include="$(PackagesDirectory)/Microsoft*.nupkg"
```
Comment (karajas) [inline build/Publish.targets]: Is there a reason that this search string changed to Microsoft.DotNet.Cli.Utils.*.nupkg?

File: build_projects/dotnet-cli-build/dotnet-cli-build.csproj [hunk 4]
```diff
@@ -29,5 +29,6 @@
     <PackageReference Include="Microsoft.Build.Framework" Version="$(CLI_MSBuild_Version)" />
     <PackageReference Include="Microsoft.DotNet.PlatformAbstractions" Version="$(PlatformAbstractionsVersion)" />
     <PackageReference Include="Microsoft.DotNet.VersionTools" Version="$(VersionToolsVersion)" />
+    <PackageReference Include="Microsoft.DotNet.Build.Tasks.Feed" Version="1.0.0-prerelease-01929-02" />
```
Comment (karajas) [inline build_projects/dotnet-cli-build/dotnet-cli-build.csproj]: This needs to be updated to latest.

File: build/publish/PublishNupkgToTransportFeed.targets [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+                AccountName="$(TransportFeedCloudDropAccountName)"
```
Comment (karajas) [inline build/publish/PublishNupkgToTransportFeed.targets]: These parameters have been updated in the latest version of the Feed Package.

**Review by karajas — CHANGES_REQUESTED** [review]
Think we should update to latest version of the package.

File: build/AzureInfo.props [hunk 6]
```diff
@@ -8,7 +8,14 @@
     <ArtifactCloudDropAccountName Condition="'$(ArtifactCloudDropAccountName)' == ''">dotnetcli</ArtifactCloudDropAccountName>
     <DotnetBlobRootUrl>https://$(ArtifactCloudDropAccountName).blob.core.windows.net/$(ArtifactContainerName)</DotnetBlobRootUrl>
     <ChecksumContainerName>$(CHECKSUM_STORAGE_CONTAINER)</ChecksumContainerName>
+    <ChecksumContainerName Condition="'$(ChecksumContainerName)' == ''">dotnet</ChecksumContainerName>
```
Comment (chcosta) [inline build/AzureInfo.props]: How is this checksum stuff related to the transport feed publishing?

File: build/AzureInfo.props [hunk 7]
```diff
@@ -8,7 +8,14 @@
...
     <ChecksumCloudDropAccessToken>$(CHECKSUM_STORAGE_KEY)</ChecksumCloudDropAccessToken>
     <ChecksumCloudDropAccountName>$(CHECKSUM_STORAGE_ACCOUNT)</ChecksumCloudDropAccountName>
+    <ChecksumCloudDropAccountName Condition="'$(ChecksumCloudDropAccountName)' == ''">dotnetclichecksums</ChecksumCloudDropAccountName>
+    <TransportFeedContainerName>$(TRANSPORTFEED_STORAGE_CONTAINER)</TransportFeedContainerName>
```
Comment (chcosta) [inline build/AzureInfo.props]: Is this pattern determining the transport feed values via environment variables?  Why not just specify them explicitly rather than relying on environment variables?  My preference is to avoid environment stuff because it makes repro's difficult (and debugging via logs)

**Comment by chcosta** [comment]
The rest of the repo's are working in master (release/2.0.0 is currently blocked for security work).  Is there a plan to enable this in Master?  According to the definition we have for "completing" phase 1, master is the requirement (though the same requirement doesn't explicitly call out CLI as a Core repo).

File: build/AzureInfo.props (hunk 7 above)
Comment (livarcocc) [reply build/AzureInfo.props]: We can't have the values in the open in our build scripts right? Otherwise anyone will be able to copy and push to our blob storage.

**Comment by livarcocc** [comment]
@chcosta this is confusing to me. My understanding from @shawnro is that all the product construction and engineering work should happen off of release/2.0.0 as that is our long term support product.

Are we saying that some of the efforts are for that branch while others are for master?

File: build_projects/dotnet-cli-build/dotnet-cli-build.csproj (hunk 4 above)
Comment (livarcocc) [reply build_projects/dotnet-cli-build/dotnet-cli-build.csproj]: Can you elaborate on what you mean by "depends on buildtools" here?

**Comment by chcosta** [comment]
Definitely an easy point of confusion.  Product construction and engineering work is *supposed* to be happening in release/2.0.0.  Most of that work is currently happening in dev/release/2.0.0 because the release/2.0.0 work is blocked on security changes.  dev/* branches don't produce official builds.  So, if we're going to make forward progress on transport feeds, we have to work in master.  Cli doesn't strictly have to be a part of that effort at this time, as there are a lot of pieces that need to be moved to transport feeds before cli can really start consuming properly.  However, for now, if there is any effort in this area, we're working in the one branch available to us, master.

**Comment by chcosta** [comment]
I pushed the requirements doc for this work to my fork this morning - https://github.com/chcosta/core-eng/blob/transport/Documentation/Project-Docs/Transport%20Feeds/requirements.md

File: build/AzureInfo.props (hunk 7 above)
Comment (chcosta) [reply build/AzureInfo.props]: They would be supplied explicitly via `/p:foo=bar`

File: build/publish/PublishNupkgToTransportFeed.targets (hunk 5 above)
Comment (johnbeisner) [reply build/publish/PublishNupkgToTransportFeed.targets]: There is just one version in the feed: https://dotnetfeed.blob.core.windows.net/dotnet-core/packages/microsoft.dotnet.build.tasks.feed/1.0.0-prerelease-01929-02/

File: build/AzureInfo.props (hunk 7 above)
Comment (johnbeisner) [reply build/AzureInfo.props]: They indeed are supplied explicitly via properties thru the VSO build definition. Please see the initial comments...

File: build_projects/dotnet-cli-build/dotnet-cli-build.csproj (hunk 4 above)
Comment (johnbeisner) [reply build_projects/dotnet-cli-build/dotnet-cli-build.csproj]: We are indeed on the latest version, there is just one version in the feed: https://dotnetfeed.blob.core.windows.net/dotnet-core/packages/microsoft.dotnet.build.tasks.feed/1.0.0-prerelease-01929-02/

File: build/Publish.targets (hunk 3 above)
Comment (johnbeisner) [reply build/Publish.targets]: To make it more explicit - or more exacting - less chance of the RegExp 'net' capturing an incorrect package.

File: build/AzureInfo.props (hunk 6 above)
Comment (johnbeisner) [reply build/AzureInfo.props]: It is not related...

File: build/AzureInfo.props (hunk 7 above)
Comment (chcosta) [reply build/AzureInfo.props]: But why pass them as environment variables, why not just use the property directly?

Comment (johnbeisner) [reply build/AzureInfo.props]: We are not passing them as environment variables. From: https://devdiv.visualstudio.com/DevDiv/_apps/hub/ms.vss-ciworkflow.build-ci-hub?_a=edit-build-definition&id=6786

Here are the args to 'build.cmd' in the "Publish Build Artifacts" step:
/p:Configuration=$(BuildConfiguration) /t:Publish  /p:ARTIFACT_STORAGE_KEY=$(ARTIFACT_STORAGE_KEY) /p:CHECKSUM_STORAGE_KEY=$(CHECKSUM_STORAGE_KEY) /p:GITHUB_PASSWORD=$(GITHUB_PASSWORD)

Comment (chcosta) [reply build/AzureInfo.props]: Ah, I guess that was the piece I was missing.  I'm still confused by the level of indirection though.  ie, why not `/p:TransportFeedContainerName=$(TRANSPORTFEED_STORAGE_CONTAINER)`?

Comment (johnbeisner) [reply build/AzureInfo.props]: "TransportFeedContainerName" is not a secret and we have a default value of "dotnet-core". If someone [like me for example] wants to test a different location, like: "jbeisner" - then the command line would have: /p:TransportFeedContainerName=jbeisner  I have done exactly this and it seems work well: https://dotnetfeed.blob.core.windows.net/jbeisner/packages/microsoft.dotnet.cli.utils/2.0.3-servicing-007047/microsoft.dotnet.cli.utils.2.0.3-servicing-007047.nupkg

Comment (chcosta) [reply build/AzureInfo.props]: I'm still not understanding the indirection, to me it just complicates things (ever so slightly).  In my mind `/p:TransportFeedContainerName=jbeisner` and `/p:TransportFeedContainerName=$(TRANSPORT_STORAGE_CONTAINER)` would be exactly how I would expect this to work.  I'm not certain what the extra layer provides since it does nothing to really making anything more or less secret.  It only means that the scripts define it one way, and a human defines it a different way.

**Review by karajas — APPROVED** [review]
Talked offline, LGTM

**Comment by chcosta** [comment]
Did we settle on where to merge this? master vs release/2.0.0?

**Comment by johnbeisner** [comment]
We are merging this PR into release/2.0.0 - it will flow into master in about 2 days.
